### PR TITLE
[Calyx] Verify continuously assigned values

### PR DESCRIPTION
**Pull request by mortbopet**
The verification ensures that:
- At most one continuous assignment exists for any given value
- A continuously assigned wire has no assignments inside groups.

### CODE REVIEW

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 1]
```diff
@@ -731,6 +731,26 @@ LogicalResult WiresOp::verify() {
...
   }
+
+  // Verify that:
+  // - At most one continuous assignment exists for any given value
+  // - A continuously assigned wire has no assignments inside groups.
+  for (auto thisAssignment : getBody()->getOps<AssignOp>()) {
+    if (thisAssignment.guard())
+      continue;
+
+    auto dest = thisAssignment.dest();
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:742]: Here as well please.

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 2]
```diff
@@ -731,6 +731,26 @@ LogicalResult WiresOp::verify() {
       return op.emitOpError() << "with name: " << groupName
                               << " is unused in the control execution schedule";
   }
+
+  // Verify that:
+  // - At most one continuous assignment exists for any given value
+  // - A continuously assigned wire has no assignments inside groups.
+  for (auto thisAssignment : getBody()->getOps<AssignOp>()) {
+    if (thisAssignment.guard())
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:739]: Maybe a comment here explaining this, e.g.,

```
// Always assume guarded assignments will not be driven simultaneously.
```

I imagine we could statically check simpler cases, such as if some wire is driven when `a` and `!a`, but I think this is fine for now.

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 3]
```diff
@@ -731,6 +731,26 @@ LogicalResult WiresOp::verify() {
...
+
+    auto dest = thisAssignment.dest();
+    for (auto *user : dest.getUsers()) {
+      auto assignUser = dyn_cast<AssignOp>(user);
+      if (!assignUser || assignUser.dest() != dest ||
+          assignUser == thisAssignment)
+        continue;
+
+      return user->emitOpError()
+             << "destination is already continuously driven";
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:750]: It might be helpful to add information here about the other assignment.

**Review by cgyurgyik — APPROVED** [review]

File: lib/Dialect/Calyx/CalyxOps.cpp (hunk 2 above)
Comment (mikeurbach) [reply lib/Dialect/Calyx/CalyxOps.cpp:739]: 👍 please add a comment and note that we are liberally (not sure if it's the right term... opposite of conservatively) assuming guards are mutually exclusive. As @cgyurgyik said there could be some interesting static checks. @rachitnigam has also suggested we could emit dynamic checks using sv.assert to ensure the property holds.

File: lib/Dialect/Calyx/CalyxOps.cpp (hunk 3 above)
Comment (mikeurbach) [reply lib/Dialect/Calyx/CalyxOps.cpp:750]: Yep, it should be easy to attachNote() and mention the other assignment.

**Review by mikeurbach — COMMENTED** [review]
Looks good, but I agree with @cgyurgyik's feedback. Also, please add a test case that two assignments to the same port are supported, if they're guarded. The low form Calyx we can import from the native compiler will have that.

File: lib/Dialect/Calyx/CalyxOps.cpp (hunk 2 above)
Comment (rachitnigam) [reply lib/Dialect/Calyx/CalyxOps.cpp:739]: Yup, the Verilator backend emits such checks and the calyx interpreter dynamically checks this property.

**Pull request merged**